nameko/concurrency/__init__.py

```diff
@@ -0,0 +1,106 @@
+"""Concurrency abstractions.
+
+Allows concurrency primitives to be abstracted such that both eventlet and
+gevent can be used to run nameko. The backend is chosen at import time and may
+be set using the global nameko config object. Valid values are "eventlet" and
+"gevent".
+
+To write nameko code and extensions that is compatible with multiple
+concurrency backend, use this compatability layer package instead of importing
+from a specific backend:
+
+NO:
+
+>>> from eventlet import sleep
+
+YES:
+
+>>> from nameko.concurrency import sleep
+
+"""
+import os
```

Comment:
I'm a bit unsure about this but here is my interpretation:

> It is important to call monkey_patch() as early in the lifetime of the application as possible. Try to do it as one of the first lines in the main module. The reason for this is that sometimes there is a class that inherits from a class that needs to be greened – e.g. a class that inherits from socket.socket – and inheritance is done at import time, so therefore the monkeypatching should happen before the derived class is defined. It’s safe to call monkey_patch multiple times.

The "os" that we are using here would be the unpatched version, but once it's patched it should be ok, and we aren't importing any other "derived" modules which would be the ones that really cause problems.  As an example:

```
import os 
print(os.open)
# <built-in function open>  # unpatched
import eventlet
eventlet.monkey_patch()
print(os.open)
# <function open at 0x7fa747035840>  # patched
import evenlet
```

I've operated with this interpretation in some of our own services and not seen any problems... However, I am not by any means an eventlet expert!!  So if you have a different interpretation I'm inclined to go with that.

Reply:
On second thought, this is probably a more general problem of having to import all of `nameko.concurrency` to import the monkey patch, since any imports within the functions that are defined in `nameko.concurrency` will be unpatched.  I think the solution needs to be for everything that is currently in the toplevel nameko.concurrency to be one level further down, so that importing the monkey_patch function can be done independently (before importing the rest), so that at the time it is called in nameko.run or whereever it is first called the rest has not been imported yet.

Reply:
Thank you for this clear explanation. I had not picked up on the subtlety only classes _derived_ from socket-importing classes being problematic for import before the monkey patch.

Agree with all your suggestions!